File: lintcode/122-Largest-Rectangle-in-Histogram/solution.java
// Brute force, O(n^n)
public class Solution {
    /**
     * @param height: A list of integer
     * @return: The area of largest rectangle in the histogram
     */
    public int largestRectangleArea(int[] height) {
        if (height == null || height.length == 0) {
            return 0;
        }
        int maxSize = 0;
        for (int i = 0; i < height.length; i++) {
            int sum = expend(i, height);
            maxSize = Math.max(maxSize, sum);
        }

        return maxSize;
    }

    private int expend(int index, int[] height) {
        int sum = height[index];
        for (int i = index - 1; i >= 0 && height[i] >= height[index]; i--) {
            sum += height[i];
        }

        for (int i = index + 1; i < height.length && height[i] >= height[index]; i++) {
            sum += height[i];
        }

        return sum;
    }
}

File: lintcode/122-Largest-Rectangle-in-Histogram/solution.cpp
class Solution {
public:
    /**
     * @param height: A list of integer
     * @return: The area of largest rectangle in the histogram
     */
    int largestRectangleArea(vector<int> &height) {
        if (height.size() == 0) {
            return 0;
        }
        
        stack<int> s;
        int best = 0;
        
        for (int i = 0; i <= height.size(); i++) {
            int curr = (i == height.size()) ? -1 : height[i];
            while (!s.empty() && curr <= height[s.top()]) {
                int h = height[s.top()];
                s.pop();
                int w = s.empty() ? i : i - s.top() - 1;
                best = max(best, h * w);
            }
            s.push(i);
        }
        
        return best;
    }
};

